components/search/src/filter.rs

```diff
@@ -0,0 +1,204 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{
+    error::Error, JSONEngineBase, JSONEngineRecord, JSONEngineUrl, JSONEngineUrls,
+    JSONSearchConfigurationRecords, RefinedSearchConfig, SearchEngineDefinition, SearchEngineUrl,
+    SearchEngineUrls, SearchUserEnvironment,
+};
+
+impl From<JSONEngineUrl> for SearchEngineUrl {
+    fn from(url: JSONEngineUrl) -> Self {
+        Self {
+            base: url.base,
+            method: url.method.unwrap_or_default().as_str().to_string(),
+            params: url.params.unwrap_or_default(),
+            search_term_param_name: url.search_term_param_name,
+        }
+    }
+}
+
+impl From<JSONEngineUrls> for SearchEngineUrls {
+    fn from(urls: JSONEngineUrls) -> Self {
+        Self {
+            search: urls.search.into(),
+            suggestions: None,
+            trending: None,
+        }
+    }
+}
+
+impl SearchEngineDefinition {
+    pub(crate) fn from_configuration_details(
+        identifier: &str,
+        base: JSONEngineBase,
+    ) -> SearchEngineDefinition {
+        SearchEngineDefinition {
+            aliases: base.aliases.unwrap_or_default(),
+            charset: base.charset.unwrap_or("UTF-8".to_string()),
+            classification: base.classification,
+            identifier: identifier.to_string(),
+            name: base.name,
+            order_hint: None,
+            partner_code: base.partner_code.unwrap_or_default(),
+            telemetry_suffix: None,
+            urls: base.urls.into(),
+        }
+    }
+}
+
+pub(crate) fn filter_engine_configuration(
+    user_environment: SearchUserEnvironment,
+    configuration: Vec<JSONSearchConfigurationRecords>,
+) -> Result<RefinedSearchConfig, Error> {
+    let mut engines = Vec::new();
+    let mut default_engine_id: Option<String> = None;
+    let mut default_private_engine_id: Option<String> = None;
+
+    for record in configuration {
+        match record {
+            JSONSearchConfigurationRecords::Engine(engine) => {
+                let result = extract_engine_config(&user_environment, engine);
+                engines.extend(result);
+            }
+            JSONSearchConfigurationRecords::DefaultEngines(default_engines) => {
+                default_engine_id = Some(default_engines.global_default);
+                default_private_engine_id.clone_from(&default_engines.global_default_private);
```

Comment:
Oooh, I haven't seen a use of `.clone_from()` in the wild before! Did Clippy recommend this?